### DISCUSSION TITLE
How to reduce more than 5 consecutive empty (=white) pixel rows to 5 white pixel rows?

### DISCUSSION THREAD

**Discussion by tobwz** [ImageMagick/ImageMagick#3678] [Help]
Assume I have a *.png snapshot image file which contains text with big space gaps between paragraphs.

I want to shrink these big gaps to acceptable space gaps.

How can I use ImageMagick to reduce more than 5 consecutive white pixel rows to 5 white pixel rows for all occurencies in this *.png file?

**Comment by snibgo** [comment]
What version of IM, on what platform?

Please show a sample input file.

Your output image will be the same width, but with lesser height, than the input. The output will be appended crops from the input. This needs a shell script. The data for the crops might come from the input scaled to a single column.

**Comment by tobwz** [comment]
v7.0.11-7 on Windows Q16 x64

sample see attached below.

>Your output image will be the same width, but with lesser height, than the input

Yes, that is correct and intended

sample image:

![How to Turn On Safe Mode in Android](https://user-images.githubusercontent.com/60912225/118285271-92439800-b4d1-11eb-9682-f72700c53008.png)

**Comment by snibgo** [comment]
The following solution is messy. There may be a better solution. [EDIT: Yes, there is a better solution. See next post.]

It is also slow, as we run "magick" for every series of white or non-white lines, and gradually build the output. If the speed matters, there are faster ways to do this.

Windows BAT script:
```
set SRC=phoneText.png

set OUTFILE=out.png

del x.txt

for /F "usebackq skip=2 tokens=3,5,9 delims=:x+, " %%A in (`%IMG7%magick ^
  %SRC% ^
  -scale "1x^!" ^
  +write info: ^
  -fill Black +opaque White ^
  -define "connected-components:verbose=true" -connected-components 4 ^
  x.png`) do (
    set HT=%%A
    set OffsY=000000%%B

    set OffsY=!OffsY:~-6!


    set Col=1
    if "%%C"=="srgb(0" set Col=0
    echo %%A %%B %%C !HT! !OffsY! !Col!
    echo !OffsY! !HT! !Col! >>x.txt
  )

rem Next "sort" is for a Unix-style sort:
rem sort <x.txt >x1.txt

rem Next "sort" is for Windows-style sort:
C:\Windows\System32\sort.exe x.txt /O x1.txt

set AppendIt=

for /F "tokens=1-3" %%A in (x1.txt) do (
  echo %%A,%%B,%%C

  if %%C==0 (
    %IMG7%magick ^
      %SRC% ^
      -crop x%%B+0+%%A +repage ^
      !AppendIt! ^
      %OUTFILE%
  ) else (
    set PadHt=%%B
    if %%B GTR 5 set PadHt=5
    %IMG7%magick ^
      -size 1x!PadHt! xc:White ^
      !AppendIt! ^
      %OUTFILE%
  )
  set AppendIt=-background Blue %OUTFILE% +swap -append

)
```
The output contains blue horizontal lines. You want white. I have made them blue because this will help you to understand what happens, and to help me if the output isn't what you want. When it does what you want, change near the bottom from...
```
  set AppendIt=-background Blue %OUTFILE% +swap -append
```
to:
```
  set AppendIt=-background White %OUTFILE% +swap -append
```
![out](https://user-images.githubusercontent.com/33812368/118321654-cd57c280-b4f5-11eb-94a3-95b908fdffee.png)

**Comment by snibgo** [comment]
An alternative strategy: after getting the data in x.txt, sort that in reverse order. Use this to build a series of "-chop" operations that reduce more than 5 white rows to only 5 rows. Store the operations in an environment variable or in a file such as a IM v7 script file. We start chopping at the bottom of the image and work upwards, hence the reverse sort.

Finally, read the source input, run the chops, and write the output.

This is much faster than my script above.
```
setlocal enabledelayedexpansion

set SRC=phoneText.png

set OUTFILE=out.png

echo off
(
for /F "usebackq skip=2 tokens=3,5,9 delims=:x+, " %%A in (`%IMG7%magick ^
  %SRC% ^
  -scale "1x^!" ^
  +write info: ^
  -fill Black +opaque White ^
  -define "connected-components:verbose=true" -connected-components 4 ^
  NULL:`) do (
    set HT=%%A
    set OffsY=000000%%B

    set OffsY=!OffsY:~-6!

    set Col=1
    if "%%C"=="srgb(0" set Col=0
    echo !OffsY! !HT! !Col!
  )
) >x.txt

rem Next "sort" is for a Unix-style sort:
rem sort --reverse <x.txt >x1.txt

rem Next "sort" is for Windows-style sort:
C:\Windows\System32\sort.exe /R x.txt /O x2.txt

set SCRFILE=ptChops.im

echo -read %SRC% >%SCRFILE%

(
for /F "tokens=1-3" %%A in (x2.txt) do (
  if %%C==1 if %%B GTR 5 (
    set /A nChop=%%B-5
    echo -chop x!nChop!+0+%%A
  )
)
) >>%SCRFILE%

echo -write %OUTFILE%  >>%SCRFILE%

%IMG7%magick ^
  -script %SCRFILE%
```
![out](https://user-images.githubusercontent.com/33812368/118371200-61cc2e80-b5a3-11eb-93fa-2bd7a99aaf07.png)

**Comment by tobwz** [comment]
@snibgo:

Thank you very much for your script.

However it is disappointing, that such a script is necessary to perform this task.

I would have expected that ImageMagick offers already a built-in function similar to

convert myimage.png -color #FFFFFF -shrinkpixelrows * 5 myimagenew.png

It would help in other scenarios as well (cropping borders, compact text with huge line heights....).

I am not sure if you are one of the developers.
Maybe this function can be added in the next release

Thank you
Toby

**Comment by snibgo** [comment]
I am not a regular IM developer, though some of my code has been incorporated.  IM is open source, and anyone can contribute.

The script takes under 5 seconds on my laptop. It wouldn't be much faster if it was built in to IM.
